File: solutions/python/stack/maximum-width-ramp.py
# LeetCode 962 — Maximum Width Ramp (Medium)
# Category: Stack · Approach: Monotonic Stack
# Time: O(n) | Space: O(n)
# Source: https://leetcode.com/problems/maximum-width-ramp/

def max_width_ramp(nums: list[int]) -> int:
    stack: list[int] = []
    for i, x in enumerate(nums):
        if not stack or nums[stack[-1]] > x:
            stack.append(i)
    res = 0
    for j in range(len(nums) - 1, -1, -1):
        while stack and nums[stack[-1]] <= nums[j]:
            res = max(res, j - stack.pop())
    return res

File: solutions/typescript/stack/maximum-width-ramp.ts
// LeetCode 962 — Maximum Width Ramp (Medium)
// Category: Stack · Approach: Monotonic Stack
// Time: O(n) | Space: O(n)
// Source: https://leetcode.com/problems/maximum-width-ramp/

function maxWidthRamp(nums: number[]): number {
  const stack: number[] = [];
  for (let i = 0; i < nums.length; i++) {
    if (stack.length === 0 || nums[stack[stack.length - 1]] > nums[i]) {
      stack.push(i);
    }
  }
  let res = 0;
  for (let j = nums.length - 1; j >= 0; j--) {
    while (stack.length && nums[stack[stack.length - 1]] <= nums[j]) {
      res = Math.max(res, j - stack.pop()!);
    }
  }
  return res;
}
